src/xdp-utils.c

```diff
@@ -572,6 +573,75 @@ cleanup_temp_file (void *p)
   g_free (*pp);
 }
 
+#define VALIDATOR_INPUT_FD 3
+
+static GKeyFile *
+spawn_validator (const char* const* args,
+                 int                fd)
+{
+  g_autoptr(GError) error = NULL;
+  g_autoptr(GBytes) bytes_output = NULL;
+  g_autoptr(GKeyFile) key_file = NULL;
+  g_autoptr(GSubprocessLauncher) launcher = NULL;
+  g_autoptr(GMappedFile) mapped_output = NULL;
+  g_autoptr(GSubprocess) process = NULL;
+  xdp_autofd int input_fd = -1;
+  int output_fd = -1;
+  xdp_autofd int stdout_fd = -1;
+
+  stdout_fd = memfd_create("spawn-validator-stdout", 0);
+
+  if (stdout_fd == -1)
+    {
+      int saved_errno;
+
+      saved_errno = errno;
+      g_warning ("Failed to spawn subprocess: memfd_create: %s", g_strerror (saved_errno));
+      return NULL;
+    }
+
+  input_fd = dup(fd);
+  // We don't need to dup() it since it will life till the launcher is dropped
+  output_fd = stdout_fd;
+
+  /* Ensure that we read from the beginning of the file */
+  lseek(input_fd, 0, SEEK_SET);
+
+  launcher = g_subprocess_launcher_new (G_SUBPROCESS_FLAGS_NONE);
+  g_subprocess_launcher_take_stdout_fd (launcher, xdp_steal_fd (&stdout_fd));
```

Comment:
This is transferring ownership. At that point using the fd (output_fd or stdout_fd) is an error! If you want to be able to use the fd while the function takes ownership, you have to dup the fd.

Nothing here is crashing because the launcher keeps the fd around until it is disposed via the autoptr but conceptually, the launcher owns the fd and could close it at any time.